### ISSUE TITLE
Pi-hole IP not reachable with macvlan configured in Fedora 32

### ISSUE THREAD

**Issue by ceesiebo** [pi-hole/docker-pi-hole#633]
I installed pi-hole version 5.0 with Docker using a macvlan network. I can see it has an IP but when I use that IP in a browser or I try to use that IP with an nslookup to find pi.hole it does not respond as if the IP does not exist.

This is a: Run Issue  

## Details

My host-IP is 192.168.178.110 and my network card is called eno1 (checked with ifconfig). I created a macvlan network with this command:

docker network create --driver=macvlan --gateway=192.168.178.1 --subnet=192.168.178.0/24 --ip-range=192.168.178.150/32 -o parent=eno1 pihole-network

After that I created the container with the docker command below.

Both ran without issues. When I check the log of the pi-hole container I see no real issues and everything started. Container is healthy according to Portainer and when I use the console and enter ip addr I see that 192.168.178.150 is returned. However, when I enter this IP in a browser, it cannot find it. When I do an nslookup pi.hole 192.168.178.150 I get a timeout and a message that this server is not reachable. When I use network bridge I can get it to work but I have to solve some port conflicts and this macvlan network seems to be the answer to that.

It is as if the firewall completely blocks access to this IP, but I turned off the firewall and the issue was still there. I also turned selinux to permissive, but that also caused no effect.

In the Docker manual I found a statement that the network card must be in promiscuous mode, so I turned that on. No effect. I found an issue that looks like mine but that was reported on a Raspberry Pi and different OS but that turned out to be a kernel defect: https://github.com/raspberrypi/linux/issues/3470. I hope that is not the case with Fedora 32.

## Related Issues
- [ ] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

## How to reproduce the issue 

1. Environment data
  * Operating System: Fedora 32 
  * Hardware: PC Intel NUC
  * Kernel Architecture: x86/amd64
  * Docker Install Info and version: version 19.03.11 build 42e35e61f3 
    - Software source: official docker-ce 
    - Supplimentary Software:portainer 
  * Hardware architecture: x86

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here

docker run -d --name pihole --net=pihole-network --ip=192.168.178.150 -p 53:53/tcp -p 53:53/udp -p 80:80 -p 443:443 -e TZ="Europe/Amsterdam" --dns=127.0.0.1 --dns=1.1.1.1 -e DNS1="8.8.8.8" -e DNS2="8.8.4.4" -v "/home/user/DockerVolumes/pihole/:/etc/pihole/:z" -v "/home/user/DockerVolumes/dnsmasq/:/etc/dnsmasq.d/:z" --restart=unless-stopped --hostname pi.hole pihole/pihole:latest

3. any additional info to help reproduce

## These common fixes didn't work for my issue

- [x ] I have tried removing/destroying my container, and re-creating a new container
- [ x] I have tried fresh volume data by backing up and moving/removing the old volume data
- [ ]x I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [ ] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [ ] I have tried running without my volume data mounts to eliminate volumes as the cause

If the above debugging / fixes revealed any new information note it here.
Add any other debugging steps you've taken or theories on root cause that may help.

**Comment by hidaba** [comment]
also in raspbian stretch

**Comment by diginc** [comment]
How does a non pi-hole container on your pi-hole macvlan compare?  Ports work fine if you run say the [default nginx container port 80](https://hub.docker.com/_/nginx/)?  Doing that eliminates the host level / mac vlan level before we dive into the pi-hole image for debugging.

Run command looks good for the most part but --ip seems redundant if you only have one IP in the macvlan range.

**Comment by buckaroogeek** [comment]
I have a docker-compose.yaml file at https://github.com/buckaroogeek/unifi-pihole-doh-docker that uses a macvlan for a pihole (and other containers).  A docker macvlan is, by default, not reachable from the host computer unless additional steps are taken with the host network configuration. Please see https://blog.oddbit.com/post/2018-03-12-using-docker-macvlan-networks/ for more information. The github repo has a small script I run on my docker host (synology NAS) that implements these changes. See the shim.sh file in the scripts directory. 

A few notes/thoughts on the docker run command.  Since this is a macvlan and the container is visible to the network directly, the ports are not mapped but exposed. So I think -p 80/tcp for example should suffice.  Also, (not sure if this makes a difference), on the volume configuration, I have always used :Z (not :z) for my Fedora docker and podman volume mounts (see for example https://community.ui.com/questions/How-To-Unifi-Controller-and-Podman/c2ca3cbc-3317-4454-995d-41f2156f6d7b for a docker/podman run command for a unifi controller).

Hope this helps.

Brad

**Comment by ceesiebo** [comment]
@diginc, basically the same problem. So it is not entirely a pi-hole problem but more a docker-macvlan-fedoraversion thing. Only for other containers I can easily use bridge mode and configure ports. To solve the port conflicts macvlan seems the easiest.

**Comment by ceesiebo** [comment]
@buckaroogeek , thanks for the links. I will try it out this evening and put my findings in the issue (and hopefully close the ticket :-)).

**Comment by diginc** [comment]
maybe try a larger ip space on the macvlan, I've never dealt with one with just one IP so that is the one thing that seems out of place to me.

Try at least a range of four like `--ip-range=192.168.178.148/30` should contain your original IP of 150

**Comment by ceesiebo** [comment]
I'm not able to create a new macvlan interface on the host with the ip link add instruction. I get an this error: Error: Attribute failed policy validation

Probably a selinux issue. I will have to check with a linux expert from work, maybe he knows. I googled but did not find an answer.

**Comment by ceesiebo** [comment]
I talked with my colleague. The RTNETLINK message is displayed because Docker has not enough privileges. So the --cap-add=NET_ADMIN is needed to be used with docker run and the ip command (also described in Docker manual https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities).

I followed instructions in the link of buckaroogeek by choosing an IP range that is not used by my DHCP on the router and create this network:

docker network create -d macvlan -o parent=eno1 --subnet=192.168.178.0/24 --gateway=192.168.178.1 --ip-range=192.168.178.201/29 --aux-address 'host=192.168.178.208' mymcvl

So I have 8 IP's at my disposal. When I now use the command ip link add mymcvl-shim link eno1 type macvlan mode bridge I get the error RTNETLINK answers: Operation not permitted. Which is logical since I did not do anything yet with the Docker run command so also not setting privileges. If I run this command:

docker run -d --name pihole --net=mymcvl --cap-add=NET_ADMIN --ip=192.168.178.201 -p 53:53/tcp -p 53:53/udp -p 80:80/tcp -p 443:443/tcp -e TZ="Europe/Amsterdam" --dns=127.0.0.1 --dns=1.1.1.1 -e DNS1="8.8.8.8" -e DNS2="8.8.4.4" -v "/home/user/DockerVolumes/pihole/:/etc/pihole/:Z" -v "/home/user/DockerVolumes/dnsmasq/:/etc/dnsmasq.d/:Z" --hostname pi.hole pihole/pihole:latest ip link add mymcvl-shim link eno1 type macvlan mode bridge

the error with RTNETLINK is gone. However, it still fails. I see in the logs that pi-hole starts successfully but when the ip command is run, the error is that it cannot find device eno1 and shuts down the container. Why can't it find device eno1? It is there. I'm able to create a network with parent=eno1.

By the way, when I try to put the ip commands in a script like buckaroogeek did I'm not able to execute the script in the docker run command. It says it cannot find the script. Does the script need to be on one of the Docker volumes?

**Comment by ceesiebo** [comment]
I found a tip and got it working! I needed to turn on promiscuous mode on my network interface eno1 (somehow it was turned off again) with : 

sudo ip link set dev eno1 promisc on

From that moment on I could enter the ip commands:

sudo ip link add shim link eno1 type macvlan mode bridge
sudo ip addr add 192.168.178.208/32 dev shim
sudo ip link set shim up
sudo ip route add 192.168.178.201 dev shim

When I use sudo ip route add 192.168.178.201/29 dev shim I get an error : invalid prefix for given prefix length. Not sure what that is about but I can add the route without the /29.

After this I started the pi-hole container (without any ip command but using the --cap-add=NET_ADMIN) and it runs perfectly. I'm able to start 192.168.178.201 in a browser and see the pi.hole screen. I'm also able to ping to the container from the host. It does not find pi.hole but that I can add the name and ip to the /etc/hosts file. Now I need to get everything configured so it also works after reboot.

Thanks all for your support.
